### PR TITLE
Add support of OS Secrets Engine

### PR DESCRIPTION
**Pull request by siyer-corp**
### Description

This PR adds Terraform provider support for Vault’s OS Secrets Engine, including backend, host, and account resources. It covers the core configuration flow for mounting the engine, registering SSH targets, and managing rotated account credentials.

- Adds `vault_os_secret_backend` to manage OS Secrets Engine backend configuration, including settings such as ssh_host_key_trust_on_first_use and backend-level config updates.

- Adds `vault_os_secret_backend_host` to manage target hosts under the OS Secrets Engine, including address, port, and host-specific rotation defaults supported by Vault.

- Adds `vault_os_secret_backend_account` to manage accounts on registered hosts, including onboarding credentials, optional rotation settings, and readback of rotation metadata such as last_vault_rotation, next_vault_rotation.

### Checklist
- [x] Added [CHANGELOG](https://github.com/hashicorp/terraform-provider-vault/blob/master/CHANGELOG.md) entry (only for user-facing changes)
- [x] Acceptance tests where run against all supported Vault Versions

### Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'

...
```
image

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

## PCI review checklist

- [ ] I have documented a clear reason for, and description of, the change I am making.

- [ ] If applicable, I've documented a plan to revert these changes if they require more than reverting the pull request.

- [ ] If applicable, I've documented the impact of any changes to security controls.

  Examples of changes to security controls include using new access control methods, adding or removing logging pipelines, etc.

### CODE REVIEW

File: internal/vault/secrets/os/os_secret_backend_resource.go [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+}
+
+func (r *OSSecretBackendResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_os_secret_backend"
+}
+
+func (r *OSSecretBackendResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Attributes: map[string]schema.Attribute{
+			consts.FieldPath: schema.StringAttribute{
```
Comment (vijayavelsekar) [inline internal/vault/secrets/os/os_secret_backend_resource.go]: This has to be `mount` field.

File: website/docs/r/os_secret_backend.html.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+### Advanced Configuration
+
+```hcl
+resource "vault_mount" "os" {
+  path = "os-prod"
+  type = "vault-plugin-secrets-os"
+}
+
+resource "vault_os_secret_backend" "os" {
+  path                             = vault_mount.os.path
```
Comment (vijayavelsekar) [inline website/docs/r/os_secret_backend.html.md]: Same here. This has to be renamed as `mount` through out this document.

File: internal/vault/secrets/os/os_secret_backend_resource_test.go [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+		Steps: []resource.TestStep{
+			{
+				Config: testAccOSSecretBackendConfig_basic(path),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, consts.FieldPath, path),
+					resource.TestCheckResourceAttr(resourceName, "max_versions", "5"),
+					resource.TestCheckResourceAttr(resourceName, "ssh_host_key_trust_on_first_use", "true"),
+				),
+			},
+			// TODO: Fix import test - currently fails because resource is destroyed between steps
```
Comment (vijayavelsekar) [inline internal/vault/secrets/os/os_secret_backend_resource_test.go]: Is this issue fixed ? are we good to enable import step now ?

File: internal/vault/secrets/os/os_secret_backend_account_resource.go [hunk 4]
```diff
@@ -0,0 +1,509 @@
...
+	_, err = cli.Logical().WriteWithContext(ctx, path, requestData)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			fmt.Sprintf("Error writing OS secret backend account to %q", path),
+			err.Error(),
+		)
+		return
+	}
+
+	// Set the ID
```
Comment (vijayavelsekar) [inline internal/vault/secrets/os/os_secret_backend_account_resource.go]: Do we need this comment ?

File: internal/vault/secrets/os/os_secret_backend_account_resource.go [hunk 5]
```diff
@@ -0,0 +1,509 @@
...
+		return
+	}
+
+	mount := data.Mount.ValueString()
+	host := data.Host.ValueString()
+	name := data.Name.ValueString()
+	path := fmt.Sprintf("%s/hosts/%s/accounts/%s", mount, host, name)
+
+	// Build the request data
+	requestData := make(map[string]interface{})
```
Comment (vijayavelsekar) [inline internal/vault/secrets/os/os_secret_backend_account_resource.go:248]: I see that we can reuse the requestData population and request to vault in Create and Update function.

File: internal/vault/secrets/os/os_secret_backend_account_resource.go [hunk 6]
```diff
@@ -0,0 +1,509 @@
...
+	_, err = cli.Logical().DeleteWithContext(ctx, path)
+	if err != nil && !util.Is404(err) {
+		resp.Diagnostics.AddError(
+			fmt.Sprintf("Error deleting OS secret backend account at %q", path),
+			err.Error(),
+		)
+	}
+}
+
+func (r *OSSecretBackendAccountResource) ImportState(ctx context.Context, req resource.ImportStateRequest, resp *resource.ImportStateResponse) {
```
Comment (vijayavelsekar) [inline internal/vault/secrets/os/os_secret_backend_account_resource.go:454]: This ImportState implementation differs from our usual approach. Can we align it with the standard pattern so the code remains concise and consistent?

Also, since the import operation will automatically invoke the Read function, we may not need to call it explicitly here. In fact, doing so could result in the Read being executed twice—once within this function and again after it completes.

Lastly, we should ensure that this works correctly for resources within a namespace as well.

File: internal/vault/secrets/os/os_secret_backend_account_resource_test.go [hunk 7]
```diff
@@ -0,0 +1,539 @@
...
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, consts.FieldMount, mount),
+					resource.TestCheckResourceAttr(resourceName, consts.FieldHost, hostName),
+					resource.TestCheckResourceAttr(resourceName, consts.FieldName, accountName),
+					resource.TestCheckResourceAttr(resourceName, consts.FieldUsername, username),
+					resource.TestCheckResourceAttr(resourceName, "rotation_period", "172800"),
+					resource.TestCheckNoResourceAttr(resourceName, "rotation_window"),
+					resource.TestCheckResourceAttr(resourceName, consts.FieldVerifyConnection, "false"),
+				),
+			},
```
Comment (vijayavelsekar) [inline internal/vault/secrets/os/os_secret_backend_account_resource_test.go:81]: we can have some import test step as well for coverage within ImportState function.

File: internal/vault/secrets/os/os_secret_backend_account_resource.go [hunk 8]
```diff
@@ -0,0 +1,509 @@
+// Copyright IBM Corp. 2016, 2026
+// SPDX-License-Identifier: MPL-2.0
+
+package os
+
+import (
+	"context"
+	"fmt"
```
Comment (vijayavelsekar) [inline internal/vault/secrets/os/os_secret_backend_account_resource.go:8]: It would be better to use `tflog`

File: internal/vault/secrets/os/os_secret_backend_host_resource.go [hunk 9]
```diff
@@ -0,0 +1,406 @@
+// Copyright IBM Corp. 2016, 2026
+// SPDX-License-Identifier: MPL-2.0
+
+package os
+
+import (
+	"context"
+	"fmt"
```
Comment (vijayavelsekar) [inline internal/vault/secrets/os/os_secret_backend_host_resource.go:8]: Good to have `tflog`

File: internal/vault/secrets/os/os_secret_backend_host_resource.go [hunk 10]
```diff
@@ -0,0 +1,406 @@
...
+		resp.Diagnostics.AddError(errutil.ClientConfigureErr(err))
+		return
+	}
+
+	mount := data.Mount.ValueString()
+	name := data.Name.ValueString()
+	path := fmt.Sprintf("%s/hosts/%s", mount, name)
+
+	// Build the request data
+	requestData := make(map[string]interface{})
```
Comment (vijayavelsekar) [inline internal/vault/secrets/os/os_secret_backend_host_resource.go]: Can we reuse requestData poplulation and writing to vault be reused between this function and Create ?

File: internal/vault/secrets/os/os_secret_backend_host_resource.go [hunk 11]
```diff
@@ -0,0 +1,406 @@
...
+	data.Mount = types.StringValue(mount)
+	data.Name = types.StringValue(name)
+
+	cli, err := client.GetClient(ctx, r.Meta(), data.Namespace.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError(errutil.ClientConfigureErr(err))
+		return
+	}
+
+	r.readHostFromVault(ctx, cli, &data, &resp.Diagnostics)
```
Comment (vijayavelsekar) [inline internal/vault/secrets/os/os_secret_backend_host_resource.go]: Should we manually call the readHostFromVault ? because after this function, it will be called anyway. Can we try our regular approach ?

File: internal/vault/secrets/os/os_secret_backend_resource.go [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+				Computed:            true,
+				Default:             booldefault.StaticBool(false),
+			},
+		},
+		MarkdownDescription: "Manages the configuration of an existing OS Secrets Engine mount in Vault.",
+	}
+	base.MustAddBaseSchema(&resp.Schema)
+}
+
+func (r *OSSecretBackendResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
```
Comment (vijayavelsekar) [inline internal/vault/secrets/os/os_secret_backend_resource.go:159]: Opportunities for reusing code for read and write operations.

File: internal/vault/secrets/os/os_secret_backend_resource.go [hunk 13]
```diff
@@ -0,0 +1,338 @@
...
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func (r *OSSecretBackendResource) Delete(_ context.Context, _ resource.DeleteRequest, _ *resource.DeleteResponse) {
+}
+
+func (r *OSSecretBackendResource) ImportState(ctx context.Context, req resource.ImportStateRequest, resp *resource.ImportStateResponse) {
+	// Use the built-in helper to set the path attribute from the import ID
+	resource.ImportStatePassthroughID(ctx, path.Root(consts.FieldPath), req, resp)
```
Comment (vijayavelsekar) [inline internal/vault/secrets/os/os_secret_backend_resource.go]: Does this implementation work for resources created inside a namespace ?

File: website/docs/r/os_secret_backend.html.md [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+* `max_versions` - (Optional, Computed) The maximum number of versions to keep. When omitted, Vault applies its server-side default of `10`. If you later remove the field from configuration, Vault retains the current value. Set to `0` to explicitly store zero in Vault.
+
+* `ssh_host_key_trust_on_first_use` - (Optional) If `true`, SSH host keys will be trusted on first use (TOFU). If `false`, host keys must be explicitly configured. Defaults to `false`.
+
+## Import
+
+OS Secret backend can be imported using the `path`, e.g.
+
+```
+$ terraform import vault_os_secret_backend.os os
```
Comment (vijayavelsekar) [inline website/docs/r/os_secret_backend.html.md:94]: we need to check if this works with TERRAFORM_VAULT_NAMESPACE_IMPORT

File: website/docs/r/os_secret_backend_account.html.md [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+### Account with Rotation Configuration
+
+```hcl
+resource "vault_mount" "os" {
+  path = "os"
+  type = "vault-plugin-secrets-os"
+}
+
+resource "vault_os_secret_backend" "os" {
+  path = vault_mount.os.path
```
Comment (vijayavelsekar) [inline website/docs/r/os_secret_backend_account.html.md]: This one also has to be `mount` and has to be fixed through out all the examples

File: website/docs/r/os_secret_backend_host.html.md [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+### Basic Host Configuration
+
+```hcl
+resource "vault_mount" "os" {
+  path = "os"
+  type = "vault-plugin-secrets-os"
+}
+
+resource "vault_os_secret_backend" "os" {
+  path = vault_mount.os.path
```
Comment (vijayavelsekar) [inline website/docs/r/os_secret_backend_host.html.md]: Here as well `mount` and in all the examples

File: internal/vault/secrets/os/os_secret_backend_account_resource_test.go (hunk 7 above)
Comment (siyer-corp) [reply internal/vault/secrets/os/os_secret_backend_account_resource_test.go:81]: ImportState is being covered in all resources

File: website/docs/r/os_secret_backend.html.md (hunk 14 above)
Comment (siyer-corp) [reply website/docs/r/os_secret_backend.html.md:94]: This is tested manually

File: website/docs/r/os_secret_backend_account.html.md [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+### Basic Account Configuration
+
+```hcl
+resource "vault_mount" "os" {
+  path = "os"
+  type = "vault-plugin-secrets-os"
+}
+
+resource "vault_os_secret_backend" "os" {
+  path = vault_mount.os.path
```
Comment (vijayavelsekar) [inline website/docs/r/os_secret_backend_account.html.md]: This field has to be `mount`.

File: website/docs/r/os_secret_backend_account.html.md (hunk 15 above)
Comment (vijayavelsekar) [inline website/docs/r/os_secret_backend_account.html.md]: This field has to be `mount`.

File: website/docs/r/os_secret_backend_account.html.md [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+### Account with Scheduled Rotation
+
+```hcl
+resource "vault_mount" "os" {
+  path = "os"
+  type = "vault-plugin-secrets-os"
+}
+
+resource "vault_os_secret_backend" "os" {
+  path = vault_mount.os.path
```
Comment (vijayavelsekar) [inline website/docs/r/os_secret_backend_account.html.md]: This field has to be `mount`.

File: website/docs/r/os_secret_backend_host.html.md (hunk 16 above)
Comment (vijayavelsekar) [inline website/docs/r/os_secret_backend_host.html.md]: This field has to be `mount`.

File: website/docs/r/os_secret_backend_host.html.md [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+### Advanced Host Configuration
+
+```hcl
+resource "vault_mount" "os" {
+  path = "os"
+  type = "vault-plugin-secrets-os"
+}
+
+resource "vault_os_secret_backend" "os" {
+  path = vault_mount.os.path
```
Comment (vijayavelsekar) [inline website/docs/r/os_secret_backend_host.html.md]: This field has to be `mount`.

File: website/docs/r/os_secret_backend_host.html.md [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+### Host with SSH Host Key
+
+```hcl
+resource "vault_mount" "os" {
+  path = "os"
+  type = "vault-plugin-secrets-os"
+}
+
+resource "vault_os_secret_backend" "os" {
+  path = vault_mount.os.path
```
Comment (vijayavelsekar) [inline website/docs/r/os_secret_backend_host.html.md]: This field has to be `mount`.

File: internal/vault/secrets/os/os_secret_backend_account_resource.go [hunk 21]
```diff
@@ -0,0 +1,514 @@
...
+	resp.Diagnostics.Append(resp.State.SetAttribute(ctx, path.Root(consts.FieldHost), host)...)
+	resp.Diagnostics.Append(resp.State.SetAttribute(ctx, path.Root(consts.FieldName), name)...)
+
+	// Password must be set after import since it's write-only
+	// Set it to a placeholder that will force user to update
+	resp.Diagnostics.Append(resp.State.SetAttribute(ctx, path.Root(consts.FieldPasswordWO), "IMPORT_PLACEHOLDER_UPDATE_REQUIRED")...)
+
+	ns := os.Getenv(consts.EnvVarVaultNamespaceImport)
+	if ns != "" {
+		tflog.Info(ctx,
```
Comment (vijayavelsekar) [inline internal/vault/secrets/os/os_secret_backend_account_resource.go:471]: nit: good to have tflog - Info/Debug in other lifecycle methods as well.

File: internal/vault/secrets/os/os_secret_backend_host_resource.go [hunk 22]
```diff
@@ -0,0 +1,385 @@
...
+		)
+		return
+	}
+
+	resp.Diagnostics.Append(resp.State.SetAttribute(ctx, path.Root(consts.FieldMount), mount)...)
+	resp.Diagnostics.Append(resp.State.SetAttribute(ctx, path.Root(consts.FieldName), name)...)
+
+	ns := os.Getenv(consts.EnvVarVaultNamespaceImport)
+	if ns != "" {
+		tflog.Info(ctx,
```
Comment (vijayavelsekar) [inline internal/vault/secrets/os/os_secret_backend_host_resource.go:409]: nit: good to have tflog - Info/Debug in other lifecycle methods as well.

File: internal/vault/secrets/os/os_secret_backend_resource.go [hunk 23]
```diff
@@ -0,0 +1,322 @@
...
+}
+
+func (r *OSSecretBackendResource) ImportState(ctx context.Context, req resource.ImportStateRequest, resp *resource.ImportStateResponse) {
+	mount := req.ID
+
+	resp.Diagnostics.Append(resp.State.SetAttribute(ctx, path.Root(consts.FieldMount), mount)...)
+
+	ns := os.Getenv(consts.EnvVarVaultNamespaceImport)
+	if ns != "" {
+		tflog.Info(ctx,
```
Comment (vijayavelsekar) [inline internal/vault/secrets/os/os_secret_backend_resource.go:307]: nit: good to have tflog - Info/Debug in other lifecycle methods as well.

File: internal/vault/secrets/os/helpers.go [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+import (
+	"fmt"
+	"regexp"
+)
+
+var (
+	// hostIDRe matches: {mount}/hosts/{name}
+	hostIDRe = regexp.MustCompile(`^([^/]+)/hosts/([^/]+)$`)
+	// accountIDRe matches: {mount}/hosts/{host}/accounts/{name}
+	accountIDRe = regexp.MustCompile(`^([^/]+)/hosts/([^/]+)/accounts/([^/]+)$`)
```
Comment (kpcraig) [inline internal/vault/secrets/os/helpers.go]: We're guilty of writing a regex for this on the plugin side too, but in addition to being different to the one in vault (e.g., https://github.com/hashicorp/vault/blob/f1fc9d3fc0175ea93d821564329accd6c8f84b27/sdk/framework/path.go#L20 and https://github.com/hashicorp/vault-plugin-secrets-os/blob/3e2e9afbc7697f79be6d7c56b335b278b8a95b14/path_accounts_rotate.go#L22) do we have any tests to convince ourselves we're capturing what we expect in all cases?

Also, is this the usual paradigm for extracting this information? (It might be, I'm not a TFVP expert)

Comment (kpcraig) [reply internal/vault/secrets/os/helpers.go]: @vinay-gopalan might have some insight here?

Comment (siyer-corp) [reply internal/vault/secrets/os/helpers.go]: @kpcraig Thanks for the callout. I see the parsing on Vault side is stricter than `[^/]+.` I notice other resources use more lenient patterns ([example](https://github.com/hashicorp/terraform-provider-vault/blob/main/vault/resource_database_secret_backend_static_role.go#L20)).

We could update the host and account name parsing to match Vault's GenericNameRegex pattern `(\w(?:(?:[\w-.]+)?\w)?)` for consistency. This would catch invalid names during import before calling Vault's API, providing clearer error messages. The mount path validation would remain unchanged?

I can update the pattern and add test coverage if you think this alignment makes sense.

Comment (vinay-gopalan) [reply internal/vault/secrets/os/helpers.go]: I took a closer look at the logic. I wanted to see if it was possible to avoid regexes to get the relevant info during imports. It looks like we will still need the regexes for the account and host resources. We won't need one for the backend resource, as we can just use the mount (currently being done in your code so that's good)

Since we can't avoid the regexes, I would lean towards Kay's suggestion, to map these regexes as close to Vault as possible. We have used lenient patterns in the past like you mentioned above, but that has lead to edge case bugs where users input weird values that can break our flows. I would say mirroring as close to Vault and adding some test coverage on the regexes should give us confidence.

On a side note, what is the idea behind setting the write-only field to state on an [import here](https://github.com/hashicorp/terraform-provider-vault/pull/2865/changes#diff-a89fac664904e42a77c60c5104dfa735b788e208e31f3cfa2a691cbc0c08f268R471)? Write-only fields shouldn't ever be written to state, even when placeholders as that is not what they are meant for. Is this line required?

Comment (siyer-corp) [reply internal/vault/secrets/os/helpers.go]: Thanks @vinay-gopalan for the review! I've updated the regex to closely match with Vault. Great catch on the Write-only field in the import, that was an inadvertent AI addition. That's removed now. Let me know if you have any other concerns.

File: .github/workflows/build.yml [hunk 25]
```diff
@@ -301,6 +322,23 @@ jobs:
           restore-keys: |
             ${{ runner.os }}-golang-
 
+      - name: Register OS plugin in Vault
```
Comment (fairclothjm) [inline .github/workflows/build.yml:325]: We are going to need this PR to be merged before we can ship the OS Secrets engine resources to production: https://github.com/hashicorp/terraform-provider-vault/pull/2707

That said, should we prioritize that PR so it can be used in this one instead of this step?

**Review by vijayavelsekar — APPROVED** [review]
LGTM
